ChrPlasmaLens element: Support k<0, and fix uninitialized-variable bug in tracking of reference particle

kyrsjo opened the pull request:
This patch adds support for negative strength (defocusing) for `ChrPlasmaLens` and `ConstF` elements. It also fixes a bug (use of uninitialized variable) in reference particle tracking for `ChrPlasmaLens`.

kyrsjo committed:
Compiles?

kyrsjo committed:
Fixed the math, no we can defocus with k<0!

kyrsjo committed:
Initial add of defocusing ChrPlasmaLens

kyrsjo committed:
Fix use of uninitialized variable for reference particle tracking in ChrPlasmaLens

kyrsjo committed:
Remove forgotten debugging comment

kyrsjo committed:
Splelling

kyrsjo commented:
Hm, as soon as I made the PR I realized I forgot to fix tabs / spaces. However the bot seems to have messed up the number of spaces needed.

I'll fix (and probably force push to override its broken fixes).

Looks like the CI is running the tests for me, that's nice :)

kyrsjo committed:
Fix tabs/spaces

kyrsjo committed:
Fix uninitialized variable giving bad results for envelope tracking with Quad when k=0

kyrsjo committed:
Handle m_g=0 correctly in ChrPlasmaLens

ax3l commented:
Thank you for your first PR to ImpactX, @kyrsjo!

Those changes look great, let us structure & test them to make them best documented/portable/covered by CI.
Please simply push additional commits as needed to the PR to update it as needed.

I also started CI and [sent you an invite to our GitHub org](https://github.com/BLAST-ImpactX), please accept to auto-run CI even before your first PR got merged on updates.

Please make sure to provide a [detailed description on the change](https://github.com/BLAST-ImpactX/impactx/pull/1030#issue-3201891770) and try to do one change at a time. e.g., if you fix a bug do it in one PR, adding a feature in another. Ideally, if it affects multiple elements as in your case, with unrelated bugs, then please make a PR per element (plasma lens, quad and ConstF here could be each a PR.)
This makes it easier for us to document changes for users and we can also "backport" fixes to older releases, when needed. It also enables us to merge things faster, because they are less complex and orthogonal changes can progress in parallel.

Please feel free to add your name to the author list of the respective files you changed.

If a specific bug is fixed (or feature added), then I usually structure my PR (branch) like this:
- a commit that adds a test that triggers the exact bug (to cover it) / covers the new feature
- another commit that adds the fix / feature
- documentation updates, if needed (usually for new features)

Welcome again, looking forward to merging those! Please let us know if you have any questions :sparkles:

ax3l commented on src/elements/ChrPlasmaLens.H:
```diff
@@ -152,21 +152,43 @@ namespace impactx::elements
-            yout = cos_ods * y + sin_ods / (omega * delta1) * py;
-            pyout = -omega * delta1 * sin_ods * y + cos_ods * py;
-
+            if (m_g > 0)
```
Just to use the right literal type:
```suggestion
            if (m_g > 0_prt)
```

kyrsjo replied:
Done on my changes.

Should this really be applied to every single litteral in the code? E.g. also here:
https://github.com/BLAST-ImpactX/impactx/pull/1030/files#diff-3858396255f74cba0b31e856f293d63c9ee64f702310a4466d58a0c1c8cfe451R63 , in ConstF constructor argument list, which includes the code
```
amrex::ParticleReal dx = 0,
amrex::ParticleReal dy = 0,
amrex::ParticleReal rotation_degree = 0,
amrex::ParticleReal aperture_x = 0,
amrex::ParticleReal aperture_y = 0,
```

ax3l replied:
For immediate assignments it does not matter that much, especially for zeros.

It is more problematic when passed into math functions and math terms, which causes doubly up and down-casting when we compile for single precision.

I just checked, and comparison to zero is actually no problem as well: https://godbolt.org/z/qYn8rfK8v

Sorry for the unnecessary comment!

ax3l commented on src/elements/ChrPlasmaLens.H:
```diff
@@ -152,21 +152,43 @@ namespace impactx::elements
+                pyout = -omega * delta1 * sin_ods * y + cos_ods * py;
+
+            }
+            else if (m_g < 0)
```
```suggestion
            else if (m_g < 0_prt)
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
-            m_const_y = -m_ky * sin_kyds;
+            // Note: The convention here is m_kx[1/m] = sign(k)*sqrt(abs(k))
+            // where k[1/m^2] is in the usual MADX convention
+            if (m_kx > 0)
```
```suggestion
            if (m_kx > 0_prt)
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
+              m_const_x  = -m_kx * sin_kxds;
+              m_sincx    = sin_kxds / m_kx;
+            }
+            else if (m_kx < 0)
```
```suggestion
            else if (m_kx < 0_prt)
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
+              m_sincx    = slice_ds;
+            }
+
+            if (m_ky > 0)
```
```suggestion
            if (m_ky > 0_prt)
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
+              m_const_y  = -m_ky * sin_kyds;
+              m_sincy    = sin_kyds / m_ky;
+            }
+            else if (m_ky < 0)
```
```suggestion
            else if (m_ky < 0_prt)
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
+              m_sincy    = slice_ds;
+            }
+
+            if (m_kt < 0)
```
```suggestion
            if (m_kt < 0_prt)
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
+
+            if (m_kt < 0)
+            {
+              throw std::runtime_error(std::string(type) + ": must have kt >= 0!");
```
We do not automatically enforce it, but just FYI: We use four spaces for indentation :)

kyrsjo replied:
Oops. Missed one!

kyrsjo replied:
Your tabs-to-spaces autoconverter did something else though, I think!

( I realized I hadn't fixed my editor config after reinstalling the laptop when your CI yelled at me and grumpily tried to to fixed it :P)

ax3l replied:
Oh no! I thought it replaces it to four spaces each :)

Visually, it might depend what you see on what your editor was configured to show for tabs before conversion?

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
+            // where k[1/m^2] is in the usual MADX convention
+
+            amrex::ParticleReal sin_kxds, cos_kxds, const_x, sincx;
+            if (m_kx > 0) {
```
```suggestion
            if (m_kx > 0_prt) {
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
+              const_x = -m_kx * sin_kxds;
+              sincx   = sin_kxds / m_kx;
+            }
+            else if (m_kx < 0) {
```
```suggestion
            else if (m_kx < 0_prt) {
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
+            }
+            else { //m_kx == 0
+              cos_kxds = 1.0;
+              const_x  = 0.0;
```
```suggestion
              cos_kxds = 1_prt;
              const_x  = 0_prt;
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
+            }
+
+            amrex::ParticleReal sin_kyds, cos_kyds, const_y, sincy;
+            if (m_ky > 0) {
```
```suggestion
            if (m_ky > 0_prt) {
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
+              const_y = -m_ky * sin_kyds;
+              sincy   = sin_kyds / m_ky;
+            }
+            else if (m_ky < 0) {
```
```suggestion
            else if (m_ky < 0_prt) {
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
+            }
+            else { //m_ky == 0
+              cos_kyds = 1.0;
+              const_y  = 0.0;
```
```suggestion
              cos_kyds = 1_prt;
              const_y  = 0_prt;
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
+              sincy    = slice_ds;
+            }
+
+            if (m_kt < 0) {
```
```suggestion
            if (m_kt < 0_prt) {
```

ax3l commented on src/elements/ConstF.H:
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
 
             // assign linear map matrix elements
             Map6x6 R = Map6x6::Identity();
+
```
Not really needed change, but ok to leave in, looks sensible.

ax3l commented on src/elements/Quad.H:
```diff
@@ -271,9 +271,9 @@ namespace impactx::elements
-                R(1,2) = m_slice_ds;
-                R(3,4) = m_slice_ds;
-                R(5,6) = m_slice_ds / betgam2;
+                R(1,2) = slice_ds;
```
Quad transport map fix for slicing: Please in a separate PR.

ax3l replied:
Merged in #1038. Please rebase against `development` or merge the latest `development` :)

kyrsjo replied:
Will do!

ax3l commented on src/elements/ChrPlasmaLens.H:
```diff
@@ -152,21 +152,43 @@ namespace impactx::elements
+                auto const cosh_ods = std::cosh(omega*m_slice_ds);
+
+                // advance transverse position and momentum (defocusing)
+                xout = cosh_ods * x + sinh_ods / (omega * delta1) * px;
```
The logic for the `R(1,1)`, `R(1,2)` etc branches here looks very similar for positive and negative values of `m_g`.

It might be worth assigning the intermediate values to variables `R_11`, `R_12`, etc., and write this branch-less (via ternary operators)/with shorter branches to:
- avoid code duplication
- reduce register pressure on GPUs

kyrsjo replied:
Sure, I was assuming that the compiler would smooth it out anyway.

And putting them in blocks like this meant I could use both const, and the `amrex::sincos` combined call (while sinh/cosh needs two calls). For the actual code, the position formulas are essentially identical, but the px has a different sign.

So it will be a lot of branches, and anyway the k=0 case is different. So if you insist I will rewrite with ternaries, but I do think it's better like this.

cemitch99 replied:
This type of branching is essentially the same as what we use in the linear Quad element.

cemitch99 replied:
A possible alternative to branching (using ternary operators) appears in the linear map of the CFbend element, line 262 [here](https://github.com/BLAST-ImpactX/impactx/blob/development/src/elements/CFbend.H).

ax3l replied:
> Sure, I was assuming that the compiler would smooth it out anyway.

If we are lucky, it could find the duplication. But I think because we do not yet use fast-math, the subtle differences in signs will prevent it.

Also no problem to leave for now, we can refactor it later.

kyrsjo replied:
OK, let's leave it for now then.

kyrsjo commented:
Hi @ax3l , thanks for the invite and for looking over it! Most fixes are easy, will do. For the "use ternary operator" thing, please see my comments.

I guess I should add tests for negative-k cases of these elements also.

Is it OK that we keep the elements in 1 PR? I'll separate out the `m_slice_ds` uninitialized stuff to a separate PR.

I see that the `apochromat_pl` and `apochromat_pl.py` tests are failing, will look into that.

General comment: I appreciate that you are insisting on small and readable PRs! Having been involved in a few codes myself, especially being the former release manager and git-repo-czar of sixtrack, it's a very good thing.

kyrsjo committed:
Use right litteral type in comparisons

kyrsjo committed:
Fix indentation

kyrsjo committed:
litteral types for assignments

ax3l commented:
Thanks a lot @kyrsjo !

> Is it OK that we keep the elements in 1 PR? I'll separate out the m_slice_ds uninitialized stuff to a separate PR.

I would really prefer to do one PR for each, because that way we can merge the quad right away and then decide for ConstF and plasma lens if and how much testing we add, too. Also makes the [release notes that we auto-generate](https://github.com/BLAST-ImpactX/impactx/releases) readable.

I already moved the Quad fix into #1038. Please merge in `development` here and it will vanish from the diff in this PR.

Please do the same for ConstF :pray:

cemitch99 reviewed (commented):
I made some small comments, and will try to help pin down why the CI tests are currently failing.

kyrsjo commented on src/elements/ChrPlasmaLens.H:
```diff
@@ -232,7 +254,7 @@ namespace impactx::elements
 
             // advance integrated path length
-            refpart.s = s + m_slice_ds;
+            refpart.s = s + slice_ds;
```
Also move this to a separate PR?

ax3l replied:
I would suggest to leave the plasma lens here and move the `ConstF` in a separate PR.

kyrsjo replied:
Done.

kyrsjo committed:
Merge branch 'development' into APL-negK

kyrsjo committed:
Rollback the changes in ConstF to what it looks in development branch; these changes will be handled in another PR.

kyrsjo commented:
OK, done. The changes to ConstF i put into PR #1041 and removed from this branch (just copied the contents of the `ConstF.h` file from the current development branch on top of my changes).

I also merged from upstream/development (before I split off the ConstF changes.)

Thanks a lot for the feedback - its really needed to get into the way of doing things in a new-to-me project!

cemitch99 commented:
@kyrsjo :  The problem is with the test `apochromatic_pl`.  If normalized focusing strength units are used (units = 0), then the test works.  The test also works if positrons are used instead of electrons.

The test inputs were impacted by PR[#935](https://github.com/BLAST-ImpactX/impactx/pull/935), and not yet updated.  The sign of the field did not matter previously, since it was assumed that k>0.  After this PR, the sign is important.  To use electrons, the sign of each focusing field (when unit=1) therefore needs to be changed (<0).

I can't push to your branch - can you modify the signs used for k in `input_apochromatic_pl.in` and `run_apochromatic_pl.py push` to this PR?  That should fix.

kyrsjo committed:
Merge branch 'development' into APL-negK

kyrsjo committed:
Fix apochromatic_pl test as suggested by @cemitch99

kyrsjo commented:
Thanks for debugging! I got a bit confused for a while since the `.in` file seems to be ignored by ctest (e.g. `ctest --test-dir build -R apochromat_pl --output-on-failure`), but after also finding the Python run file, it worked.

kyrsjo commented:
@cemitch99 weird that you cannot push to branch btw - ` Allow edits by maintainers ` is on. I'll just add you both to the fork repo.

Thanks again for the feedback! Sorry that I'm slow these days - it's holiday time.

kyrsjo commented:
I see there are some conflicts after the changes in PR #1002 - will fix in branch.

kyrsjo committed:
Merge branch 'development' into APL-negK
(Fixed merge conflict & manually added the removal of std:: for sinh/cosh)

ax3l commented on examples/apochromatic/input_apochromatic_pl.in:
```diff
@@ -35,43 +35,43 @@ dr2.ds = 2.0
 q1.type = plasma_lens_chromatic
 q1.ds = 0.331817852986604588
-q1.k = 996.147787384348956
+q1.k = -996.147787384348956
```
Maybe I am just confused, byt why are all the lenses now set to defocusing (or are they)?

Do we still cover the positive k case in CI somewhere?

cemitch99 replied:
@ax3l Since `q1.units = 1`, the field is given in SI units.   Since these are electrons, the focusing field is negative.

ax3l replied:
But the unit was not changed... was this a bug before the PR?

kyrsjo replied:
We have `T_Real const omega = sqrt(abs(m_g)/delta1);` and `m_g = m_unit == 1 ? m_k / refpart.rigidity_Tm() : m_k;`.

In the old code version `omega` is always used with focusing matrix elements, whatever the sign of `m_g` or `k`. So the sign of `m_g` was effectively discarded, and it was forced to be a focusing element.

kyrsjo commented:
FYI I'm working on a test for a defocusing lens.

Is there a test for a lens that's off?

kyrsjo committed:
Chromatic effect in drift as well, as in ChrQuad and in ChrDrift

kyrsjo committed:
Merge branch 'development' into APL-negK

kyrsjo committed:
Fix time propagation for g=0 in ChrPlasmaLens

kyrsjo committed:
Fix time dependency for negative focusing

kyrsjo commented on src/elements/ChrPlasmaLens.H:
```diff
@@ -265,7 +318,9 @@ namespace impactx::elements
         // see: compute_constants() to refresh
         amrex::ParticleReal m_slice_ds;  //! m_ds / nslice();
         amrex::ParticleReal m_beta;
+        amrex::ParticleReal m_gamma;
```
Should these be `T_Real` as well? I'm not really sure about the difference vs `amrex::ParticleReal`.

This method gets called pre-tracking, and the variables get used in the tracking calculations.

ax3l replied:
No, because `m_gamma` and others calculated in `compute_constants` are independent of individual particles (they might at best depend on the reference particle).

We use `T_Real` only in the per-particle push to enable SIMD vectorization (pushing multiple particles at once). When we use `-DImpactX_SIMD=ON`, the types behind `T_Real` will do independent, point-wise math (roughly: with an overloaded `double[N]` or `float[N]`) instead of scalar math (`amrex::ParticleReal` aka `double`/`float`).

kyrsjo committed:
Merge branch 'development' into APL-negK

kyrsjo committed:
Merge branch 'development' into APL-negK

kyrsjo committed:
Start adding tests for APLs

kyrsjo committed:
Activate the APL test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Refactoring

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Improve ChrPlasmaLens tests and add plot

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Test the ChrPlasmaLens, not drift

kyrsjo committed:
Coding conventions

kyrsjo committed:
Cleanup

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Add focusing test to ChrPlasmaLens

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Add defocusing test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Cleanup and add to CMakeLists

kyrsjo commented on examples/active_plasma_lens/analysis_APL_ChrPlasmaLens_defocusing.py:
```diff
@@ -0,0 +1,80 @@
+)
+
+atol = 0.0  # ignored
+# rtol = 2.2 * num_particles**-0.5  # from random sampling of a smooth distribution
```
Where does this 2.2 actually come from? For some reason, the sample sigma seems to always start a bit high, and with defocusing this is blown up to *just* over the limit.

I relaxed the constrain to let it pass, but I'm not sure if it's the right thing to do.

<img width="1238" height="918" alt="image" src="https://github.com/user-attachments/assets/7483651d-209e-450d-9f93-7c0a3751d0a3" />

cemitch99 replied:
The 2.2 factor is rather arbitrary.  We know the statistical fluctuations in the moments due to finite N should scale as sqrt(N), and this proportionality factor is chosen largely by trial-and-error.  (We start large, so that the test passes, and then decrease this value so that tests pass barely.)  Relaxing the constraint is just fine.

ax3l replied:
Exactly. Please feel free to slighly relax the constant from 2.2 to a higher number if need be.

kyrsjo commented on examples/active_plasma_lens/run_APL_ChrPlasmaLens.py:
```diff
@@ -0,0 +1,164 @@
+
+def analytic_final_estimate(APL_g, rigidity_Tm, APL_length, beta_0, alpha_0):
+    ## Note: this seems to not work correctly for k != 0???
+    k = APL_g * rigidity_Tm
```
This should be `/` not `*`

kyrsjo replied:
Fixed.

kyrsjo committed:
Fix analytical estimate, generalize run script for test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Fix error in CMakeLists

kyrsjo committed:
Rename run_APL_ChrPlasmaLens.py -> run_APL.py

kyrsjo committed:
Swap reference values to analytical values

kyrsjo committed:
Catch error

kyrsjo committed:
Add README.rst for active_plasma_lens_test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kyrsjo committed:
Hopefully fixing REST syntax

kyrsjo commented:
I think this is ready for review/merging now, @ax3l & @cemitch99 ?

There are two questions I wonder about (datatype and tolerance) + 1 question I've answered (why negative).

Note that I've made the tests so that I can fit the ConstF tests into the same framework :)

kyrsjo committed:
Merge branch 'development' into APL-negK

kyrsjo committed:
Merge branch 'development' into APL-negK

cemitch99 committed:
Merge branch 'development' into APL-negK

cemitch99 commented on examples/active_plasma_lens/run_APL.py:
```diff
@@ -0,0 +1,197 @@
+
+    # Forward-propagate that through the focusing/defocusing lens
+    # from the beginning, ignoring energy spread
+    # [Doesn't give same result as ChrPlasmaLens and ChrQuad for some reason, even when sigpt_0 = 0]
```
Does this comment still apply?  If so, this needs to be addressed.

kyrsjo replied:
Hi, this was fixed here:

https://github.com/BLAST-ImpactX/impactx/pull/1030/commits/bb670a1f2877cb999bc6f13a30728eb7e5f62c8c#diff-a95909e2ce0d4e807a3fa1dc18e09c9c822ec1dbb1275b60b056c775609e216eL125

and then I started using the analytical estimate here:
https://github.com/BLAST-ImpactX/impactx/pull/1030/commits/7f06ab254a786cb5e946063a241fcde4dd0723a0

I will cleanup when I do constF, PR #1041 - the tests should share quite a bit of code.

cemitch99 commented on examples/active_plasma_lens/analysis_APL_ChrPlasmaLens.py:
```diff
@@ -0,0 +1,64 @@
+    beta_y = sigy2 / emittance_y
+    alpha_y = -epstrms["position_y"]["momentum_y"] / emittance_y
+
+    return (beta_x, beta_y, alpha_x, alpha_y)
```
Note:  This works, but the values defined here are computed internally and stored as reduced_diagnostics.  See this example for an alternative approach:  https://impactx.readthedocs.io/en/latest/usage/examples/fodo_space_charge/README.html
It's fine to keep this for now.

cemitch99 commented on examples/active_plasma_lens/plot_APL_ChrPlasmaLens_analytical.py:
```diff
@@ -0,0 +1,72 @@
+
+import matplotlib.pyplot as plt
+import numpy as np
+from run_APL_ChrPlasmaLens import analytic_final_estimate
```
```suggestion
from run_APL import analytic_final_estimate
```

cemitch99 replied:
Just caught this - name change.  I just tried it.

kyrsjo replied:
Thank you!

cemitch99 committed:
Update examples/active_plasma_lens/plot_APL_ChrPlasmaLens_analytical.py

cemitch99 committed:
Link README to examples documentation.

ax3l commented on docs/source/usage/examples.rst:
```diff
@@ -41,6 +41,7 @@ Single Particle Dynamics
    examples/reversibility/README.rst
    examples/charge_sign/README.rst
    examples/symplectic_integration/README.rst
+   examples/active_plasma_lens
```
```suggestion
   examples/active_plasma_lens/README.rst
```

cemitch99 replied:
```suggestion
   examples/active_plasma_lens/README.rst
```

cemitch99 committed:
Update docs/source/usage/examples.rst

ax3l reviewed (approved):
Looks great, thank you for this! :rocket: :sparkles:

cemitch99 commented on examples/active_plasma_lens/README.rst:
```diff
@@ -0,0 +1,45 @@
+Active Plasma Lens
+==================
+
+These examples demonstrates the of the effect of an Active Plasma Lens (APL) on the beam.
```
```suggestion
These examples demonstrate the effect of an Active Plasma Lens (APL) on the beam.
```

cemitch99 commented on examples/active_plasma_lens/README.rst:
```diff
@@ -0,0 +1,45 @@
+The beam is set to have :math:`\alpha = 0` in the middle of the lens in the case of no field.
+The beam size in the middle of the lens is set to 10 µm for the no-field examples (in order to have a strongly parabolic :math:`\beta`-function within the lens), and 100 µm for the focusing and defocusing examples.
+A :math:`\sigma_{pt} = 10^{-3}` is also assumed.
+Before the simulations, this beam is back-propagate to the lens entry asuming no field.
```
```suggestion
Before the simulation, this beam is back-propagated to the lens entry assuming zero field.
```

cemitch99 committed:
Apply suggestions from code review

kyrsjo commented:
Thank you @ax3l and @cemitch99 for merging! I will now proceed to ConstF. Also, PhD student @aellings will now be able to simulate diverging APLs in ImpactX :)

Maybe see you at Elba?

ax3l commented:
@kyrsjo Awesome, thanks for the contribution!!

Yes, I am at EAAC until Thursday noon, let's catch up!
